Initial work on projections

jonclare opened the pull request.

jonclare committed:
Throwing some stuff at the wall around projections

jonclare committed:
Further work on projections. First SQL projection is now working

carlosrfernandez reviewed (commented):
🚀

jonclare committed:
Clean-up around projections metadata.
Now supports mapping an event to multiple types and each type partipating in multiple contexts

jonclare committed:
Generate create table script automatically if user doesn't supply their own script

jonclare committed:
Add a simple builder for SqlColumnDefinitions

jonclare committed:
Support deletions from sql projections

jonclare committed:
Support for executing custom sql from events

daniel-smith commented on src/DomainLib.Projections.Sqlite/SqliteSqlDialect.cs:
```diff
@@ -0,0 +1,138 @@
+
+        private static string GetDataTypeName(DbType dbType)
+        {
+            switch (dbType)
```
WOAH!!!

jonclare replied:
Hahaha

daniel-smith commented on src/DomainLib.Projections.Sql/ISqlDialect.cs:
```diff
@@ -0,0 +1,14 @@
+namespace DomainLib.Projections.Sql
+{
+    public interface ISqlDialect
+    {
```
Are we able to use the interfaces here, i.e. `IDbConnection`, `IDbCommand`?

daniel-smith replied:
Also, I wonder if we're missing something else...I'm not sure that something that knows how to provide SQL in a given dialect should also be creating DB connections/commands. Just a thought. It may be too early to separate these concepts at the moment.

jonclare replied:
I thought about that too. The interface has since been renamed to IDbConnector as it's more than just a SQL dialect now.
I did wonder about separating out the part that generates SQL from the part that provides the connection/binds parameters.

Those two things are coupled in some way due to the fact that they're both related to the database type. Agree that it's not full SRP at the moment though. We can think some more about it later

jonclare replied:
I've removed some of the uses of DbCommand now, as I've changed the DbConnector to return strings for the upsert and delete commands.
That allows us to consolidate the SQL that's run if a user wants to do an upsert/delete and run some custom sql for a projection.
The rest are going to have to stay as it's only the abstract classes that provide async overloads when talking to the db

jonclare committed:
Add notification when caught up to live event stream and handle that in SqlContext. Added some parameterisation there too, so users can configure handling of transactions

jonclare committed:
Clean up, better boundary checks and ensure that we don't go back onto a consumer's synchronisation context

jonclare committed:
Allow parameter bindings to be specified by the user instead of always using the reflection based bindings.
Generate DbCommands when the builders are built rather than in the fluent interface. This allows automated and custom sql to be concatenated together
General clean up

jonclare committed:
Clean up

jonclare committed:
Add some better logging and configurable values in EventStream

jonclare committed:
Rejig some things around for easier testability.
Initial work on getting some tests going

jonclare committed:
Better tests for projections library

jonclare committed:
Add more complex test using an in-memory DB.
Fix bugs that were highlighted by this test

jonclare committed:
Better management of SqlContext lifetimes. Fixes broken tests
